lib/pacemaker/pcmk_sched_ordering.c

```diff
@@ -1208,54 +1209,66 @@ find_actions_by_task(pe_resource_t *rsc, const char *original_key)
+                             const pe_resource_t *rsc, pe__ordering_t *order)
 {
     GList *then_actions = NULL;
-    pe_action_t *then_action = NULL;
-    uint32_t flags;
+    uint32_t flags = pe_order_none;
 
-    CRM_CHECK(rsc != NULL, return);
-    CRM_CHECK(order != NULL, return);
+    CRM_CHECK((rsc != NULL) && (order != NULL), return);
 
     flags = order->flags;
-    then_action = order->rh_action;
-    crm_trace("Applying ordering constraint %d (then: %s)", order->id, rsc->id);
+    pe_rsc_trace(rsc, "Applying ordering constraint %d for 'then' resource %s",
+                 order->id, rsc->id);
 
-    if (then_action != NULL) {
-        then_actions = g_list_prepend(NULL, then_action);
+    if (order->rh_action != NULL) {
+        then_actions = g_list_prepend(NULL, order->rh_action);
 
-    } else if (rsc != NULL) {
+    } else {
         then_actions = find_actions_by_task(rsc, order->rh_action_task);
     }
 
     if (then_actions == NULL) {
-        pe_rsc_trace(rsc,
-                     "Ignoring constraint %d: then (%s for %s) not found",
+        pe_rsc_trace(rsc, "Ignoring constraint %d: no %s actions found for %s",
                      order->id, order->rh_action_task, rsc->id);
         return;
     }
 
     if ((first_action != NULL) && (first_action->rsc == rsc)
         && pcmk_is_set(first_action->flags, pe_action_dangle)) {
 
-        pe_rsc_trace(rsc, "Detected dangling operation %s -> %s",
-                     first_action->uuid, order->rh_action_task);
+        pe_rsc_trace(rsc,
+                     "Detected dangling migration ordering (%s then %s %s)",
+                     first_action->uuid, order->rh_action_task, rsc->id);
         pe__clear_order_flags(flags, pe_order_implies_then);
     }
 
-    for (GList *gIter = then_actions; gIter != NULL; gIter = gIter->next) {
-        pe_action_t *then_action_iter = (pe_action_t *) gIter->data;
+    if ((first_action == NULL) && !pcmk_is_set(flags, pe_order_implies_then)) {
+        pe_rsc_info(rsc, "Ignoring ordering %d: No first action found",
+                    order->id);
+        g_list_free(then_actions);
+        return;
+    }
+
+    for (GList *iter = then_actions; iter != NULL; iter = iter->next) {
+        pe_action_t *then_action_iter = (pe_action_t *) iter->data;
 
         if (first_action != NULL) {
             order_actions(first_action, then_action_iter, flags);
-
-        } else if (pcmk_is_set(flags, pe_order_implies_then)) {
-            pe__clear_action_flags(then_action_iter, pe_action_runnable);
-            crm_warn("Unrunnable %s %#.6x", then_action_iter->uuid, flags);
         } else {
-            crm_warn("neither %s %#.6x", then_action_iter->uuid, flags);
+            pe__clear_action_flags(then_action_iter, pe_action_runnable);
+            crm_warn("Action %s is unrunnable due to ordering %d "
+                     CRM_XS " flags=%#06" PRIx32,
+                     then_action_iter->uuid, order->id, flags);
```

Comment:
If we're improving log messages, do you think we should use something other than the ordering ID in the user-friendly (non-detail) message? It looks like the ID is an integer that gets incremented every time a new ordering constraint is created, so it's not in the CIB and doesn't seem meaningful to an end user.

Reply:
Yeah that bugged me too. Most orderings are internal rather than in the CIB so that makes it a little more challenging. I don't think it matters for debug/trace but higher level messages really should be reworded ...

Reply:
Updated commit